tokio/src/task/scope/cancellation_token.rs

```diff
@@ -0,0 +1,255 @@
+            // only move it from the blocked into the ready state and not have
+            // further side effects.
+
+            let waiters = self.waiters.take();
+
+            unsafe {
+                // Use a reverse iterator, so that the oldest waiter gets
+                // scheduled first
+                for waiter in waiters.into_reverse_iter() {
+                    if let Some(handle) = (*waiter).task.take() {
+                        handle.wake();
+                    }
+                    (*waiter).state = PollState::Done;
+                }
+            }
+        }
+    }
+
+    fn is_cancelled(&self) -> bool {
+        self.is_cancelled
+    }
+
+    /// Checks if the cancellation has occured. If it is this returns immediately.
+    /// If the event isn't set, the WaitForCancellationFuture gets added to the wait
+    /// queue at the event, and will be signalled once ready.
+    /// This function is only safe as long as the `wait_node`s address is guaranteed
+    /// to be stable until it gets removed from the queue.
+    unsafe fn try_wait(
+        &mut self,
+        wait_node: &mut ListNode<WaitQueueEntry>,
+        cx: &mut Context<'_>,
+    ) -> Poll<()> {
+        match wait_node.state {
+            PollState::New => {
+                if self.is_cancelled {
+                    // The event is already signaled
+                    wait_node.state = PollState::Done;
+                    Poll::Ready(())
+                } else {
+                    // Added the task to the wait queue
+                    wait_node.task = Some(cx.waker().clone());
+                    wait_node.state = PollState::Waiting;
+                    self.waiters.add_front(wait_node);
+                    Poll::Pending
+                }
+            }
+            PollState::Waiting => {
+                // The WaitForCancellationFuture is already in the queue.
+                // The event can't have been set, since this would change the
+                // waitstate inside the mutex.
+                // We don't need to update the Waker here, since we assume the
+                // Future is only ever polled from the same task. If this wouldn't
+                // hold, true, this wouldn't be safe.
+                Poll::Pending
+            }
+            PollState::Done => {
+                // We have been woken up by the event.
+                // This does not guarantee that the event is still set. It could
+                // have been reset it in the meantime.
+                Poll::Ready(())
+            }
+        }
+    }
+
+    fn remove_waiter(&mut self, wait_node: &mut ListNode<WaitQueueEntry>) {
+        // WaitForCancellationFuture only needs to get removed if it has been added to
+        // the wait queue of the Event. This has happened in the PollState::Waiting case.
+        if let PollState::Waiting = wait_node.state {
+            if !unsafe { self.waiters.remove(wait_node) } {
```

Comment:
In this method, we have exclusive mutable access to `self.waiters`. Naively, I would expect to be able to remove from a list with a `&mut` ref safely...i've not gotten to read the linked list implementation yet, though.

Reply:
It's indeed safe as long as the list is consistent. But the outside API of the list at this point just forwards the `unsafe` annotations that are required internally. And since it mutates a raw pointer of non-owned elements, it is unsafe.